scalac-scoverage-runtime/shared/src/main/scala/scoverage/Invoker.scala

```diff
@@ -59,6 +66,78 @@ object Invoker {
     }
   }
 
+  /**
+   * We record that the given id has been invoked by appending its id to the coverage
+   * data file.
+   *
+   * This will happen concurrently on as many threads as the application is using,
+   * so we use one file per thread, named for the thread id.
+   *
+   * This method is not thread-safe if the threads are in different JVMs, because
+   * the thread IDs may collide.
+   * You may not use `scoverage` on multiple processes in parallel without risking
+   * corruption of the measurement file.
+   *
+   * Same method as above but added additional features to make it work with
+   * https://github.com/scoverage/scalac-scoverage-plugin/issues/265. Typically, we clean the dataDir
+   * and add coverage instruments to the measurement file at runtime rather than compile time. This is
+   * needed as the runtime process may be running on a different host than the compile time one.
+   * However, since the instruments are on the classpath, we can retrieve the instruments from different hosts.
+   *
+   * @param id the id of the statement that was invoked
+   * @param instrumentsDir the directory where the instrument data is held and needs to be copied from
+   */
+  def invokedWriteToClasspath(id: Int, instrumentsDir: String): Unit = {
+    // Get the output data directory from the environment variable or write to a subdir.
+    var dataDir: String = ""
+    sys.env.get(measurementFileEnv) match {
+      case Some(d) => dataDir = d
+      case None => dataDir = subdir
+    }
+
+    // Adding subdir to the basedir. This is because every
+    // instrumented binary has a coverage object for which measurements will be generated
+    // and it is possible that a test file tests multiple instrumented binaries. Thus, to
+    // segregate reports for each source file, we need this hash. This was not an issue with
+    // previous version as then we explicitly specified a different [dataDir] each time.
+    dataDir += s"/${safe_name(instrumentsDir)}"
```

Comment:
Ditto doing this at startup time. You should probably add a `setter` for the datadir for tests to use that will do the appropriate concatenation.

This `invoke` method needs to be very fast since it will be called millions of times per run. Doing too much setup here will definitely slow down an application.

Reply:
I think this concatenation needs to be at each invocation as it is a parameter to the `invokedWriteToClasspath` method